sui_types/src/gas.rs

```diff
@@ -19,75 +25,304 @@ macro_rules! ok_or_gas_error {
+    pub object_deletion_per_byte_rebate: StorageCost,
+}
+
+// TODO: The following numbers are arbitrary at this point.
+static INIT_SUI_COST_TABLE: Lazy<SuiCostTable> = Lazy::new(|| SuiCostTable {
+    min_transaction_cost: ComputationCost(InternalGasUnits::new(10000)),
+    package_publish_per_byte_cost: ComputationCost(InternalGasUnits::new(80)),
+    object_read_per_byte_cost: ComputationCost(InternalGasUnits::new(15)),
+    object_mutation_per_byte_cost: ComputationCost(InternalGasUnits::new(40)),
+    consensus_cost: ComputationCost(InternalGasUnits::new(100000)),
+
+    new_owned_storage_per_byte_cost: StorageCost(InternalGasUnits::new(100)),
+    new_shared_storage_per_byte_cost: StorageCost(InternalGasUnits::new(500)),
+
+    object_deletion_per_byte_rebate: StorageCost(InternalGasUnits::new(70)),
+});
+
+pub static MAX_GAS_BUDGET: Lazy<u64> =
+    Lazy::new(|| u64::MAX / INITIAL_COST_SCHEDULE.gas_constants.gas_unit_scaling_factor);
+
+pub static MIN_GAS_BUDGET: Lazy<u64> = Lazy::new(|| {
+    let consts = &INITIAL_COST_SCHEDULE.gas_constants;
+    consts
+        .to_external_units(INIT_SUI_COST_TABLE.min_transaction_cost.0)
+        .get()
+});
+
+pub struct SuiGasStatus<'a> {
+    gas_status: GasStatus<'a>,
+    init_budget: u64,
+    storage_cost: InternalGasUnits<GasCarrier>,
+    storage_rebate: InternalGasUnits<GasCarrier>,
 }
 
+impl<'a> SuiGasStatus<'a> {
+    pub fn new_with_budget(gas_budget: u64) -> SuiGasStatus<'a> {
+        Self::new(
+            GasStatus::new(&INITIAL_COST_SCHEDULE, GasUnits::new(gas_budget)),
+            gas_budget,
+        )
+    }
+
+    pub fn new_unmetered() -> SuiGasStatus<'a> {
+        Self::new(GasStatus::new_unmetered(), 0)
+    }
+
+    pub fn get_move_gas_status(&mut self) -> &mut GasStatus<'a> {
+        &mut self.gas_status
+    }
+
+    pub fn charge_min_tx_gas(&mut self) -> SuiResult {
+        self.deduct_computation_cost(&INIT_SUI_COST_TABLE.min_transaction_cost)
+    }
+
+    pub fn charge_consensus(&mut self) -> SuiResult {
+        self.deduct_computation_cost(&INIT_SUI_COST_TABLE.consensus_cost)
+    }
+
+    pub fn charge_publish_package(&mut self, size: usize) -> SuiResult {
+        let computation_cost = INIT_SUI_COST_TABLE
+            .package_publish_per_byte_cost
+            .with_size(size);
+        self.deduct_computation_cost(&computation_cost)
+    }
+
+    pub fn charge_storage_read(&mut self, size: usize) -> SuiResult {
+        let cost = INIT_SUI_COST_TABLE
+            .object_read_per_byte_cost
+            .with_size(size);
+        self.deduct_computation_cost(&cost)
+    }
+
+    pub fn charge_storage_mutation(&mut self, old_size: usize, new_size: usize) -> SuiResult {
+        let cost = INIT_SUI_COST_TABLE
+            .object_mutation_per_byte_cost
+            .with_size(new_size);
+        self.deduct_computation_cost(&cost)?;
+        // When the size of the object changes, we also charge or rebate the storage cost.
```

Comment:
Might be slightly clearer to have an explicit no-op for the case where `new_size == old_size`. I think the `rebate_storage_deletion` path will end up doing the right thing, but it's not obvious.